parser/chunk.go

```diff
@@ -121,35 +137,162 @@ func (c *Chunk) Parse(r io.Reader, options *ChunkParseOptions) (err error) {
+			if _, err := c.rd.SeekStart(c.pointer); err != nil {
+				c.err = fmt.Errorf("unable to seek to position %d: %w", c.pointer, err)
+				return false
 			}
-			size, err := rd.VarInt()
+			size, err := c.rd.VarInt()
 			if err != nil {
-				return fmt.Errorf("unable to parse event size: %w", err)
+				c.err = fmt.Errorf("unable to parse event size: %w", err)
+				return false
 			}
-			events[pointer] = size
-			e, err := ParseEvent(rd, classes, cpools)
+			e, err := ParseEvent(c.rd, c.classes, c.cpools)
 			if err != nil {
-				return fmt.Errorf("unable to parse event: %w", err)
+				c.err = fmt.Errorf("unable to parse event: %w", err)
+				return false
 			}
-			c.Events = append(c.Events, e)
-			pointer += int64(size)
+			c.Event = e
+			c.pointer += int64(size)
+			return true
 		}
 	}
-	return nil
+	return false
+}
+
+// Err returns the first error encountered by Events.
+func (c *Chunk) Err() error {
+	return c.err
 }
 
-func buildClasses(metadata MetadataEvent) ClassMap {
-	classes := make(map[int]ClassMetadata)
-	for _, class := range metadata.Root.Metadata.Classes {
+func (c *Chunk) buildClasses(metadata MetadataEvent) ClassMap {
+	cacheEventFns := map[string]func() Parseable{
+		"jdk.CPULoad": func() Parseable {
+			c.cpuLoad = CPULoad{}
+			return &c.cpuLoad
+		},
+		"jdk.ThreadPark": func() Parseable {
+			c.threadPark = ThreadPark{}
+			return &c.threadPark
+		},
+		"jdk.ExecutionSample": func() Parseable {
+			c.executionSample = ExecutionSample{} // threadstate
+			return &c.executionSample
+		},
+		"jdk.ObjectAllocationInNewTLAB": func() Parseable {
+			c.objectAllocationInNewTLAB = ObjectAllocationInNewTLAB{}
+			return &c.objectAllocationInNewTLAB
+		},
+		"jdk.InitialSystemProperty": func() Parseable {
+			c.initialSystemProperty = InitialSystemProperty{}
+			return &c.initialSystemProperty
+		},
+		"jdk.ActiveSetting": func() Parseable {
+			c.activeSetting = ActiveSetting{}
+			return &c.activeSetting
+		},
+		"jdk.NativeLibrary": func() Parseable {
+			c.nativeLibrary = NativeLibrary{}
+			return &c.nativeLibrary
+		},
+	}
+	classes := make(map[int]*ClassMetadata, len(metadata.Root.Metadata.Classes))
+	for i := range metadata.Root.Metadata.Classes {
+		class := &metadata.Root.Metadata.Classes[i]
+		var numConstants int
+		for _, field := range class.Fields {
+			if field.ConstantPool {
+				numConstants++
+			}
+		}
+
+		if cacheEventFn, ok := cacheEventFns[class.Name]; ok {
+			class.eventFn = cacheEventFn
+		} else if eventFn, ok := events[class.Name]; ok {
+			class.eventFn = eventFn
+		} else {
+			class.eventFn = func() Parseable {
+				return &UnsupportedEvent{}
+			}
+		}
+
+		if typeFn, ok := types[class.Name]; ok {
+			class.typeFn = typeFn
+		} else {
+			class.typeFn = func() ParseResolvable {
+				return &UnsupportedType{}
+			}
+		}
+		class.numConstants = numConstants
 		classes[int(class.ID)] = class
 	}
+
+	parseBaseTypeAndDrops := map[string]func(reader.Reader) error{
```

Comment:
nit: can we make `parseBaseTypeAndDrops` global?

Reply:
> nit: can we make `parseBaseTypeAndDrops` global?

i do not want parseBaseTypeAndDrops global.
because if it is a global function, then we should at least two map find:
1, use class id to find class metadata.
2, use class name to find correspond basetype.